Add support for distkey.

benjyw opened the pull request:
Subject: Add support for setting a DISTKEY.

### Feature or Bugfix
- Feature

### Purpose
Setting a DISTKEY is a basic requirement in real-world Redshift clusters. 
This change adds support for designating a column as the DISTKEY.
There are some complications and caveats, explained in the README.


### Detail
This is implemented by setting a custom index type in the model's meta class:

```
from django_redshift_backend.base import DistKey
...
class Fact(Model):
  class Meta:
    indexes = [DistKey(fields=['dim1'])]

  dim1 = ForeignKey(Dimension1Model)
```

But, as explained in the README, this will likely require some hand-editing
of the migration files, so that all necessary information is available at 
table creation time.

This change also fixes a minor issue with the SORTKEY support, where
it was assumed that the db column name was the same as the field name.
But this is not the case, e.g., for ForeignKey fields.

This change also replaces references to the defunct `remote_field.to` with
the `remote_field.related_model`.

### Relates
https://github.com/shimizukawa/django-redshift-backend/issues/16
https://github.com/shimizukawa/django-redshift-backend/pull/20

shimizukawa commented:
@benjyw Thanks! Please correct flake8 error.

shimizukawa commented on README.rst:
```diff
@@ -87,6 +87,137 @@ There is built-in support for this option for Django >= 1.9. To use `sortkey`, s
+---------------------------------
+
+There is built-in support for this option for Django >= 1.11. To use `distkey`, define an index on the model
+meta with the custom index type `django_redshift_backend.base.DistKey` with `fields` naming a single field::
```
I suggest `django_redshift_backend.models.DistKey` or `django_redshift_backend.DistKey`.

Because `django_redshift_backend.base` is internal name that reflects [django.db.backends.postgres.base](https://github.com/django/django/tree/master/django/db/backends/postgresql), we should not use `base` for users.

benjyw replied:
Moved it to `django_redshift_backend.distkey.Distkey` if that's OK. `django_redshift_backend.models.DistKey` seemed wrong because this isn't a Django model. And `django_redshift_backend.DistKey` would require adding code in `__init__.py` which I prefer to keep empty. 

Does this make sense?

shimizukawa replied:
Agreed. `django_redshift_backend.DistKey` is better.

shimizukawa commented on README.rst:
```diff
@@ -87,6 +87,137 @@ There is built-in support for this option for Django >= 1.9. To use `sortkey`, s
+
+However Redshift requires that the `distkey` be specified at table creation. As a result, you may need to
+manually edit your migration files to move the ForeignKey column into the initial `CreateModel`, while also
+ensuring that the referenced table appears *before* the referencing table in the file.
```
Hum... I think editing migration files *before* migration is not a good experience. On the other hand, I feel it is useful to be able to automatically provided DISTKEY at migration phase.
@Surgo How do you think?

benjyw replied:
The migration docs explicitly say that you're welcome to edit or even hand-craft migration files (and in some extreme cases you have to). `makemigrations` is just a convenience.

Surgo replied:
This is not `ForeignKey` specific problems.

1. Get related fields like `ForeignKey` and `ManyToMany` from model
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L518-L534
1. Create table without related fields
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L557-L569
1. Add field for all related fields
    https://github.com/django/django/blob/2.0.7/django/db/migrations/autodetector.py#L575-L589

If we will not edit migration file, we have to override `generate_created_models`.

benjyw replied:
Modifying the migration generation code seemed like it would potentially introduce instability, bugs etc. It might be nice to add in the future, but it's likely outside the scope of this change.

shimizukawa replied:
No objection from me.

benjyw commented:
I'm still chasing down an issue with using multiple databases in Django. It looks like it ignores `allow_migrate()` in the custom router in various circumstances. Best to hold off on this PR until I fix that, in case the fix ends up requiring changes here (I doubt it, but you never know).

Surgo commented on django_redshift_backend/base.py:
```diff
@@ -160,8 +166,8 @@ def create_model(self, model):
             if field.remote_field and field.db_constraint:
-                to_table = field.remote_field.to._meta.db_table
-                to_column = field.remote_field.to._meta.get_field(
+                to_table = field.remote_field.related_model._meta.db_table
```
Its bug to migrate django 2.0... Sorry

https://github.com/shimizukawa/django-redshift-backend/pull/43

shimizukawa replied:
#43 has been merged. @benjyw  please rebase this PR if you have chance.

benjyw commented:
Addressed comments and fixed style errors.

benjyw commented:
Are you using django_redshift_backend in a scenario where you have multiple databases in your Django app, e.g., Redshift and a regular Postgres transactional db? I cannot seem to get this to work, as Django insists on applying all migrations on both dbs, even though I have created a router and overridden `allow_migrate()` (and verified that it is running and returning the right value). From eyeballing the Django source it looks like the router simply isn't consulted in the codepaths I'm hitting. Have you noticed anything like this?

benjyw commented:
Hi - any update on this? I think I've addressed your concerns?

benjyw commented:
Just to clarify, do you want me to move `django_redshift_backend.distkey.Distkey` to `django_redshift_backend.Distkey`, or just update the documentation, which currently points to the wrong place?

benjyw commented:
Fixed README and rebased after #43 .

Let me know if you want me to move `django_redshift_backend.distkey.DistKey` to `django_redshift_backend.DistKey`. I wasn't sure if that's what you were requesting. I generally like to keep `__init__.py` empty, but if you prefer it there, happy to move it.

shimizukawa commented:
Oh, I missed that the `DistKey` class moved from `base.py` to `distkey.py`. The document was also fixed, so this implementation is OK. I will merge this.

shimizukawa commented:
Merged. Thanks!

benjyw commented:
Thank you!
What's your release strategy? When might I expect this to be released?

shimizukawa commented:
Actually we have no release strategy, but I'll make a release within two days. #44

shimizukawa commented:
0.9.0 has been released! #44
